File: frontend/src/services/custom.ts
import { http } from './http';
import { 
    CustomIndicator, CustomIndicatorCreatePayload,
    CustomSignal, CustomSignalCreatePayload,
    CustomPosition, CustomPositionCreatePayload,
    CustomRiskRule, CustomRiskRuleCreatePayload,
    CustomComponentListResponse 
} from '../types/custom';

// ================= Indicators =================
export const getIndicators = async (params?: { skip?: number; limit?: number; category?: string }) => {
  const { data } = await http.get<CustomComponentListResponse<CustomIndicator>>('/custom/indicators', { params });
  return data;
};

export const createIndicator = async (payload: CustomIndicatorCreatePayload) => {
  const { data } = await http.post<CustomIndicator>('/custom/indicators', payload);
  return data;
};

export const updateIndicator = async (id: string, payload: Partial<CustomIndicatorCreatePayload>) => {
  const { data } = await http.put<CustomIndicator>(`/custom/indicators/${id}`, payload);
  return data;
};

export const deleteIndicator = async (id: string) => {
  const { data } = await http.delete(`/custom/indicators/${id}`);
  return data;
};

// ================= Signals =================
export const getSignals = async (params?: { skip?: number; limit?: number; category?: string }) => {
  const { data } = await http.get<CustomComponentListResponse<CustomSignal>>('/custom/signals', { params });
  return data;
};

export const createSignal = async (payload: CustomSignalCreatePayload) => {
  const { data } = await http.post<CustomSignal>('/custom/signals', payload);
  return data;
};

export const updateSignal = async (id: string, payload: Partial<CustomSignalCreatePayload>) => {
  const { data } = await http.put<CustomSignal>(`/custom/signals/${id}`, payload);
  return data;
};

export const deleteSignal = async (id: string) => {
  const { data } = await http.delete(`/custom/signals/${id}`);
  return data;
};

// ================= Positions =================
export const getPositions = async (params?: { skip?: number; limit?: number; category?: string }) => {
  const { data } = await http.get<CustomComponentListResponse<CustomPosition>>('/custom/positions', { params });
  return data;
};

export const createPosition = async (payload: CustomPositionCreatePayload) => {
  const { data } = await http.post<CustomPosition>('/custom/positions', payload);
  return data;
};

export const updatePosition = async (id: string, payload: Partial<CustomPositionCreatePayload>) => {
  const { data } = await http.put<CustomPosition>(`/custom/positions/${id}`, payload);
  return data;
};

export const deletePosition = async (id: string) => {
  const { data } = await http.delete(`/custom/positions/${id}`);
  return data;
};

// ================= Risk Rules =================
export const getRiskRules = async (params?: { skip?: number; limit?: number; category?: string }) => {
  const { data } = await http.get<CustomComponentListResponse<CustomRiskRule>>('/custom/risk-rules', { params });
  return data;
};

export const createRiskRule = async (payload: CustomRiskRuleCreatePayload) => {
  const { data } = await http.post<CustomRiskRule>('/custom/risk-rules', payload);
  return data;
};

export const updateRiskRule = async (id: string, payload: Partial<CustomRiskRuleCreatePayload>) => {
  const { data } = await http.put<CustomRiskRule>(`/custom/risk-rules/${id}`, payload);
  return data;
};

export const deleteRiskRule = async (id: string) => {
  const { data } = await http.delete(`/custom/risk-rules/${id}`);
  return data;
};

File: backend/services/custom.py
from typing import Any
import uuid
from typing import List, Optional, Type
from sqlalchemy.orm import Session
from sqlalchemy import select
from loguru import logger

from backend.models.custom import CustomIndicator, CustomSignal, CustomPosition, CustomRiskRule
from backend.schemas.custom import (
    CustomIndicatorCreate, CustomIndicatorUpdate,
    CustomSignalCreate, CustomSignalUpdate,
    CustomPositionCreate, CustomPositionUpdate,
    CustomRiskRuleCreate, CustomRiskRuleUpdate
)

class CustomComponentService:
    def __init__(self, db: Session):
        self.db = db

    # Helper method for generic CRUD
    async def _get_list(self, model: Type, skip: int = 0, limit: int = 100, category: Optional[str] = None) -> dict:
        stmt = select(model).order_by(model.updated_at.desc())
        if category:
            stmt = stmt.where(model.category == category)
            
        count_stmt = select(model.id)
        if category:
            count_stmt = count_stmt.where(model.category == category)
            
        total = len(self.db.execute(count_stmt).fetchall())
        stmt = stmt.offset(skip).limit(limit)
        items = self.db.execute(stmt).scalars().all()
        
        return {"total": total, "items": items}

    async def _get_by_id(self, model: Type, item_id: uuid.UUID):
        return self.db.execute(select(model).where(model.id == item_id)).scalar_one_or_none()

    async def _create(self, model: Type, data: Any, user_id: uuid.UUID):
        try:
            item_dict = data.model_dump()
            item_dict["author_id"] = user_id
            
            new_item = model(**item_dict)
            self.db.add(new_item)
            self.db.commit()
            self.db.refresh(new_item)
            
            logger.info(f"Created new {model.__name__}: {new_item.name}")
            return new_item
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to create {model.__name__}: {str(e)}")
            raise e

    async def _update(self, model: Type, item_id: uuid.UUID, data: Any):
        try:
            item = await self._get_by_id(model, item_id)
            if not item:
                return None
                
            update_data = data.model_dump(exclude_unset=True)
            for key, value in update_data.items():
                setattr(item, key, value)
                
            self.db.commit()
            self.db.refresh(item)
            return item
        except Exception as e:
            self.db.rollback()
            raise e

    async def _delete(self, model: Type, item_id: uuid.UUID) -> bool:
        try:
             item = await self._get_by_id(model, item_id)
             if not item:
                 return False
                 
             self.db.delete(item)
             self.db.commit()
             return True
        except Exception as e:
             self.db.rollback()
             raise e

    # ============== Custom Indicators ==============
    async def get_indicators(self, skip: int = 0, limit: int = 100, category: str = None):
        return await self._get_list(CustomIndicator, skip, limit, category)

    async def get_indicator_by_id(self, item_id: uuid.UUID):
        return await self._get_by_id(CustomIndicator, item_id)

    async def create_indicator(self, data: CustomIndicatorCreate, user_id: uuid.UUID):
        return await self._create(CustomIndicator, data, user_id)

    async def update_indicator(self, item_id: uuid.UUID, data: CustomIndicatorUpdate):
        return await self._update(CustomIndicator, item_id, data)

    async def delete_indicator(self, item_id: uuid.UUID):
        return await self._delete(CustomIndicator, item_id)
        
    # ============== Custom Signals ==============
    async def get_signals(self, skip: int = 0, limit: int = 100, category: str = None):
        return await self._get_list(CustomSignal, skip, limit, category)

    async def get_signal_by_id(self, item_id: uuid.UUID):
        return await self._get_by_id(CustomSignal, item_id)

    async def create_signal(self, data: CustomSignalCreate, user_id: uuid.UUID):
        return await self._create(CustomSignal, data, user_id)

    async def update_signal(self, item_id: uuid.UUID, data: CustomSignalUpdate):
        return await self._update(CustomSignal, item_id, data)

    async def delete_signal(self, item_id: uuid.UUID):
        return await self._delete(CustomSignal, item_id)

    # ============== Custom Positions ==============
    async def get_positions(self, skip: int = 0, limit: int = 100, category: str = None):
        return await self._get_list(CustomPosition, skip, limit, category)

    async def get_position_by_id(self, item_id: uuid.UUID):
        return await self._get_by_id(CustomPosition, item_id)

    async def create_position(self, data: CustomPositionCreate, user_id: uuid.UUID):
        return await self._create(CustomPosition, data, user_id)

    async def update_position(self, item_id: uuid.UUID, data: CustomPositionUpdate):
        return await self._update(CustomPosition, item_id, data)

    async def delete_position(self, item_id: uuid.UUID):
        return await self._delete(CustomPosition, item_id)

    # ============== Custom Risk Rules ==============
    async def get_risk_rules(self, skip: int = 0, limit: int = 100, category: str = None):
        return await self._get_list(CustomRiskRule, skip, limit, category)

    async def get_risk_rule_by_id(self, item_id: uuid.UUID):
        return await self._get_by_id(CustomRiskRule, item_id)

    async def create_risk_rule(self, data: CustomRiskRuleCreate, user_id: uuid.UUID):
        return await self._create(CustomRiskRule, data, user_id)

    async def update_risk_rule(self, item_id: uuid.UUID, data: CustomRiskRuleUpdate):
        return await self._update(CustomRiskRule, item_id, data)

    async def delete_risk_rule(self, item_id: uuid.UUID):
        return await self._delete(CustomRiskRule, item_id)
